File: test/CSharp/CodeCracker.Test/Refactoring/AllowMembersOrderingCodeFixProviderTests.StyleCop.cs
using CodeCracker.Refactoring;
using System.Threading.Tasks;
using TestHelper;
using Xunit;

namespace CodeCracker.Test.Refactoring
{
    public class StyleCopAllowMembersOrderingCodeFixProviderTests : CodeFixTest<AllowMembersOrderingAnalyzer, StyleCopAllowMembersOrderingCodeFixProvider>
    {
        [Fact]
        public async Task StyleCopAllowMembersOrderingShouldAssureMembersOrderByType()
        {
            const string source = @"
            class Foo
            {
                public class Class { }
                public struct Sruct { }
                void Method() { }
                public int this[int a] { set { } }
                public int Property { set { } }
                public interface Interface { }
                enum Enum { Enum1 }
                public event System.Action Event { add { } remove { } }
                public delegate double Delegate();
                ~Foo() { }
                Foo() { }
                public event System.Action EventField;
                public int b = 0;
            }";

            const string expected = @"
            class Foo
            {
                public int b = 0;
                Foo() { }
                ~Foo() { }
                public delegate double Delegate();
                public event System.Action EventField;
                public event System.Action Event { add { } remove { } }
                enum Enum { Enum1 }
                public interface Interface { }
                public int Property { set { } }
                public int this[int a] { set { } }
                void Method() { }
                public struct Sruct { }
                public class Class { }
            }";

            await VerifyCSharpFixAsync(source, expected);
        }

        [Fact]
        public async Task StyleCopAllowMembersOrderingShouldAssureMembersOrderByModifiers()
        {
            const string source = @"
            public class Foo
            {
                private int p = 0;
                protected int q = 0;
                protected internal int r = 0;
                internal int s = 0;
                public int t = 0;
                static int u = 0;
                public static int v = 0;
                const int x = 0;
                public const int z = 0;
            }";

            const string expected = @"
            public class Foo
            {
                public const int z = 0;
                public static int v = 0;
                public int t = 0;
                internal int s = 0;
                protected internal int r = 0;
                protected int q = 0;
                const int x = 0;
                static int u = 0;
                private int p = 0;
            }";

            await VerifyCSharpFixAsync(source, expected);
        }

        [Fact]
        public async Task StyleCopAllowMembersOrderingShouldAssureMembersOrderByAlphabeticalOrder()
        {
            const string source = @"
            class Foo
            {
                int c = 2, d = 3;
                int a = 0, b = 1;
                event System.Action EventField1;
                event System.Action EventField;
                delegate double Delegate2();
                delegate double Delegate();
                event System.Action Event1 { add { } remove { } }
                event System.Action Event { add { } remove { } }
                enum Enum1 { Enum1 }
                enum Enum { Enum1 }
                interface Interface2 { }
                interface Interface { }
                int Property1 { set { } }
                int Property { set { } }
                public static int operator +(Foo a, Foo b) { return 0; }
                public static int operator -(Foo a, Foo b) { return 0; }
                void Method1() { }
                void Method() { }
                public struct Sruct1 { }
                public struct Sruct { }
                public class Class { }
                public class Class1 { }
            }";

            const string expected = @"
            class Foo
            {
                int a = 0, b = 1;
                int c = 2, d = 3;
                delegate double Delegate();
                delegate double Delegate2();
                event System.Action EventField;
                event System.Action EventField1;
                event System.Action Event { add { } remove { } }
                event System.Action Event1 { add { } remove { } }
                enum Enum { Enum1 }
                enum Enum1 { Enum1 }
                interface Interface { }
                interface Interface2 { }
                int Property { set { } }
                int Property1 { set { } }
                public static int operator -(Foo a, Foo b) { return 0; }
                public static int operator +(Foo a, Foo b) { return 0; }
                void Method() { }
                void Method1() { }
                public struct Sruct { }
                public struct Sruct1 { }
                public class Class { }
                public class Class1 { }
            }";

            await VerifyCSharpFixAsync(source, expected);
        }

        [Theory]
        [InlineData("struct")]
        [InlineData("class")]
        public async Task StyleCopAllowMembersOrderingShouldAssureMembersOrderByStyleCopPatterns(string typeDeclaration)
        {
            var source = @"
            using System;
            namespace ConsoleApplication1
            {
                " + typeDeclaration + @" Foo
                {
                    public class Foo2 { }
                    public struct Struct { }
                    public static Foo operator +(Foo f1, Foo f2) { return new Foo(); }
                    public static Foo operator -(Foo f1, Foo f2) { return new Foo(); }
                    void Method(string a) { }
                    internal void Method(int a) { }
                    public void Method1() { }
                    public void Method() { }
                    public string this[int i] { set { } }
                    public string Property { get; set; }
                    public interface Interface { }
                    public enum Enum { Enum1, Enum2 = 1 }
                    public event Action Event
                    {
                        add { EventField += value; }
                        remove { EventField -= value; }
                    }
                    public delegate double Delegate(double num);
                    public Foo()
                    {
                        Property = Field1 = Field = "";
                        EventField = EventField1 = () => { };
                    }
                    public event Action EventField1;
                    public event Action EventField;
                    public string Field;
                    public static string Field1;
                }
            }";

            var expected = @"
            using System;
            namespace ConsoleApplication1
            {
                " + typeDeclaration + @" Foo
                {
                    public static string Field1;
                    public string Field;
                    public Foo()
                    {
                        Property = Field1 = Field = "";
                        EventField = EventField1 = () => { };
                    }
                    public delegate double Delegate(double num);
                    public event Action EventField;
                    public event Action EventField1;
                    public event Action Event
                    {
                        add { EventField += value; }
                        remove { EventField -= value; }
                    }
                    public enum Enum { Enum1, Enum2 = 1 }
                    public interface Interface { }
                    public string Property { get; set; }
                    public string this[int i] { set { } }
                    public static Foo operator -(Foo f1, Foo f2) { return new Foo(); }
                    public static Foo operator +(Foo f1, Foo f2) { return new Foo(); }
                    public void Method() { }
                    public void Method1() { }
                    internal void Method(int a) { }
                    void Method(string a) { }
                    public struct Struct { }
                    public class Foo2 { }
                }
            }";

            await VerifyCSharpFixAsync(source, expected);
        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Refactoring/AllowMembersOrderingCodeFixProviderTests.Stylecop.vb
Imports CodeCracker.VisualBasic.Refactoring
Imports Xunit

Namespace Refactoring
    Public Class StyleCopAllowMembersOrderingCodeFixProviderTests
        Inherits CodeFixTest(Of AllowMembersOrderingAnalyzer, StyleCopAllowMembersOrderingCodeFixProvider)

        <Fact>
        Public Async Function StyleCopAllowMembersOrderingShouldAssureMembersOrderingByType() As Task
            Const source = "
Class Foo
    Public Class Class1

    End Class
    Public Structure Struct

    End Structure
    Sub Method()

    End Sub
    Public Property Prop As Integer
    Public Interface IInterface

    End Interface
    Public Enum Enum1
        EnumVal
    End Enum
    Public Custom Event Event1 As System.Action
        AddHandler(value As System.Action)

        End AddHandler
        RemoveHandler(value As System.Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Public Delegate Function Delegate1() As Double
    Sub New()

    End Sub
    Public Event EventField As System.Action
    Public b = 0
End Class"

            Const expected = "
Class Foo
    Public b = 0
    Sub New()

    End Sub
    Public Delegate Function Delegate1() As Double
    Public Event EventField As System.Action
    Public Custom Event Event1 As System.Action
        AddHandler(value As System.Action)

        End AddHandler
        RemoveHandler(value As System.Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Public Enum Enum1
        EnumVal
    End Enum
    Public Interface IInterface

    End Interface
    Public Property Prop As Integer
    Sub Method()

    End Sub
    Public Structure Struct

    End Structure
    Public Class Class1

    End Class
End Class"

            Await VerifyBasicFixAsync(source, expected)
        End Function

        <Fact>
        Public Async Function StyleCopAllowMembersOrderingShouldAssureMembersOrderByModifiers() As Task
            Const source = "
Public Class Foo1
    Private p = 0
    Protected q = 0
    Protected Friend r = 0
    Friend s = 0
    Public t = 0
    Shared u = 0
    Public Shared v = 0
    Const x = 0
    Public Const z = 0
End Class"

            Const expected = "
Public Class Foo1
    Public Const z = 0
    Public Shared v = 0
    Public t = 0
    Friend s = 0
    Protected Friend r = 0
    Protected q = 0
    Const x = 0
    Shared u = 0
    Private p = 0
End Class"

            Await VerifyBasicFixAsync(source, expected)
        End Function

        <Fact>
        Public Async Function StyleCopAllowmembersOrderingShouldAssureMembersOrderByAlphabeticalOrder() As Task
            Const source = "
Public Class Foo2
    Private c = 2, d = 3
    Private a = 0, b = 1
    Event EventField1 As System.Action
    Event EventField As System.Action
    Delegate Function Delegate2() As Double
    Delegate Function Delegate1() As Double
    Custom Event Event2 As System.Action
        AddHandler(value As System.Action)

        End AddHandler
        RemoveHandler(value As System.Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Custom Event Event1 As System.Action
        AddHandler(value As Action)

        End AddHandler
        RemoveHandler(value As Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Enum Enum2
        Enum1
    End Enum
    Enum Enum1
        Enum1
    End Enum
    Interface Interface2

    End Interface
    Interface Interface1

    End Interface
    Property Property2 As Integer
    Property Property1 As Integer
    Public Shared Operator +(A As Foo2, B As Foo2)
        Return 0
    End Operator
    Public Shared Operator -(a As Foo2, b As Foo2)
        Return 0
    End Operator
    Sub Method2()

    End Sub
    Sub Method1()

    End Sub
    Public Structure Sruct1

    End Structure
    Public Structure Sruct

    End Structure
    Public Class Class2

    End Class
    Public Class Class1
    End Class
End Class"

            Const expected = "
Public Class Foo2
    Private a = 0, b = 1
    Private c = 2, d = 3
    Delegate Function Delegate1() As Double
    Delegate Function Delegate2() As Double
    Event EventField As System.Action
    Event EventField1 As System.Action
    Custom Event Event1 As System.Action
        AddHandler(value As Action)

        End AddHandler
        RemoveHandler(value As Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Custom Event Event2 As System.Action
        AddHandler(value As System.Action)

        End AddHandler
        RemoveHandler(value As System.Action)

        End RemoveHandler
        RaiseEvent()

        End RaiseEvent
    End Event
    Enum Enum1
        Enum1
    End Enum
    Enum Enum2
        Enum1
    End Enum
    Interface Interface1

    End Interface
    Interface Interface2

    End Interface
    Property Property1 As Integer
    Property Property2 As Integer
    Public Shared Operator -(a As Foo2, b As Foo2)
        Return 0
    End Operator
    Public Shared Operator +(A As Foo2, B As Foo2)
        Return 0
    End Operator
    Sub Method1()

    End Sub
    Sub Method2()

    End Sub
    Public Structure Sruct

    End Structure
    Public Structure Sruct1

    End Structure
    Public Class Class1
    End Class
    Public Class Class2

    End Class
End Class"

            Await VerifyBasicFixAsync(source, expected)
        End Function

        <Theory>
        <InlineData("Structure")>
        <InlineData("Class")>
        Public Async Function StyleCopAllowMembersOrderingShouldAssureMembersOrderByStyleCopPatterns(typeDeclaration As String) As Task
            Dim source = "
Imports System
Namespace ConsoleApplication1
    " & typeDeclaration & " Foo3
        Public Class Foo4
        End Class
        Public Structure Struct

        End Structure
        Public Shared Operator +(f1 As Foo3, f2 As Foo3)
            Return New Foo3
        End Operator
        Public Shared Operator -(f1 As Foo3, f2 As Foo3)
            Return New Foo3
        End Operator
        Sub Method(a As String)

        End Sub
        Friend Sub Method(a As Integer)

        End Sub
        Public Sub Method1()

        End Sub
        Public Sub Method()

        End Sub
        Public Property Prop As String
        Public Interface IInterface

        End Interface
        Public Enum Enum1
            Enum1
            Enum2
        End Enum
        Public Custom Event Event1 As Action
            AddHandler(value As Action)

            End AddHandler
            RemoveHandler(value As Action)

            End RemoveHandler
            RaiseEvent()

            End RaiseEvent
        End Event
        Public Delegate Function Delegate1(num As Double) As Double
        Public Sub New()
            Prop = """"
            Field1 = """"
            Field = """"
        End Sub
        Public Event EventField1 As Action
        Public Event EventField As Action
        Public Field As String
        Public Shared Field1 As String
    End " & typeDeclaration & "

End Namespace"

            Dim expected = "
Imports System
Namespace ConsoleApplication1
    " & typeDeclaration & " Foo3
        Public Shared Field1 As String
        Public Field As String
        Public Sub New()
            Prop = """"
            Field1 = """"
            Field = """"
        End Sub
        Public Delegate Function Delegate1(num As Double) As Double
        Public Event EventField As Action
        Public Event EventField1 As Action
        Public Custom Event Event1 As Action
            AddHandler(value As Action)

            End AddHandler
            RemoveHandler(value As Action)

            End RemoveHandler
            RaiseEvent()

            End RaiseEvent
        End Event
        Public Enum Enum1
            Enum1
            Enum2
        End Enum
        Public Interface IInterface

        End Interface
        Public Property Prop As String
        Public Shared Operator -(f1 As Foo3, f2 As Foo3)
            Return New Foo3
        End Operator
        Public Shared Operator +(f1 As Foo3, f2 As Foo3)
            Return New Foo3
        End Operator
        Public Sub Method()

        End Sub
        Public Sub Method1()

        End Sub
        Friend Sub Method(a As Integer)

        End Sub
        Sub Method(a As String)

        End Sub
        Public Structure Struct

        End Structure
        Public Class Foo4
        End Class
    End " & typeDeclaration & "

End Namespace"

            Await VerifyBasicFixAsync(source, expected)
        End Function
    End Class
End Namespace